### PR TITLE
Fix the request target in PSR7 Request

### PR DESCRIPTION
**Pull request by mtibben**
The request target is not being preserved correctly when using this library. This is because the query string is normalized, and we're not populating the PSR-7 request with the actual request target.

This is important for things like http signatures which uses the request target as an input to sign a request.

### CODE REVIEW

**Comment by mtibben** [comment]
I've added a fix for the issue. 

The failing tests here are also failing on master.

**Comment by xabbuh** [comment]
I opened #31 to fix the test suite.

**Comment by fabpot** [comment]
@mtibben You can now rebase to validate that tests pass now.

**Comment by mtibben** [comment]
I'm traveling over the next week without computer, feel free to rebase :)

**Comment by mtibben** [comment]
Rebased, and tests are passing @fabpot

File: Factory/DiactorosFactory.php [hunk 1]
```diff
@@ -55,7 +55,7 @@ public function createRequest(Request $symfonyRequest)
         $request = new ServerRequest(
             $server,
             DiactorosRequestFactory::normalizeFiles($this->getFiles($symfonyRequest->files->all())),
-            $symfonyRequest->getUri(),
+            $symfonyRequest->getSchemeAndHttpHost().$symfonyRequest->getRequestUri(),
```
Comment (xabbuh) [inline Factory/DiactorosFactory.php]: Is this change really required? Shouldn't the result be the same?

File: Factory/DiactorosFactory.php [hunk 2]
```diff
@@ -65,6 +65,7 @@ public function createRequest(Request $symfonyRequest)
             ->withCookieParams($symfonyRequest->cookies->all())
             ->withQueryParams($symfonyRequest->query->all())
             ->withParsedBody($symfonyRequest->request->all())
+            ->withRequestTarget($symfonyRequest->getRequestUri())
```
Comment (xabbuh) [inline Factory/DiactorosFactory.php]: Why is this necessary when we already pass the full URI to the constructor?

File: Factory/DiactorosFactory.php (hunk 1 above)
Comment (mtibben) [inline Factory/DiactorosFactory.php]: I've done this defensively to avoid repeats of the problem I have encountered.

[`getUri`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Request.php#L1118) creates a URL in a very indirect way. This will likely be a problem due to the way Symfony requests normalise different parts of the URL. In comparison, [`getRequestUri`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Request.php#L1089) is much more direct, you can see in [`prepareRequestUri`](https://github.com/symfony/symfony/blob/master/src/Symfony/Component/HttpFoundation/Request.php#L1775) how the URL comes directly from the superglobals or headers.

As we're dealing with a conversion between request types, we need to use the cleanest data we can when constructing the PSR7 request. Otherwise we can get a URL different to the original

File: Factory/DiactorosFactory.php (hunk 2 above)
Comment (mtibben) [inline Factory/DiactorosFactory.php]: Note that a HTTP request begins with a request target (e.g `GET /request-target?query=string`) **not** a URL.

In the same way, the [PSR-7 Request interface](https://github.com/php-fig/http-message/blob/master/src/RequestInterface.php) does not mention URLs or constructors. It only knows about request targets. Therefore, I think we should be explicit about setting the request target correctly using [`withRequestTarget`](https://github.com/php-fig/http-message/blob/master/src/RequestInterface.php#L61) rather than relying on a particular implementation's constructor

Comment (xabbuh) [inline Factory/DiactorosFactory.php]: But we are dealing with a concrete implementation here. And if its constructor is not working as expected (which we should check) the implementation should be fixed instead.

Comment (mtibben) [inline Factory/DiactorosFactory.php]: Yep I understand what you're saying, I was thinking the same myself. Yet when I looked at the implementation I thought it best to be explicit here. This implementation tokenizes the URL, and pieces together a request target if it is not set explicitly. By providing the request target string, you're giving the implementation the correct data - unfortunately the constructor is not giving you that opportunity.  https://github.com/zendframework/zend-diactoros/blob/master/src/RequestTrait.php#L124-L129

Comment (mtibben) [inline Factory/DiactorosFactory.php]: In fact tests fail without this line, exactly for this reason
```
Testing Symfony PSR-7 HTTP message Bridge Test Suite
F............                                                     13 / 13 (100%)

Time: 89 ms, Memory: 4.00MB

There was 1 failure:

1) Symfony\Bridge\PsrHttpMessage\Tests\Factory\DiactorosFactoryTest::testCreateRequest
Failed asserting that two strings are equal.
--- Expected
+++ Actual
@@ @@
-'/testCreateRequest?foo=1&bar[baz]=42'
+'/testCreateRequest?foo=1&bar%5Bbaz%5D=42'

/Users/michael/src/github.com/mtibben/psr-http-message-bridge/Tests/Factory/DiactorosFactoryTest.php:87
/Users/michael/src/github.com/mtibben/psr-http-message-bridge/vendor/symfony/phpunit-bridge/bin/.phpunit/phpunit-5.7/phpunit:5

FAILURES!
Tests: 13, Assertions: 63, Failures: 1.
```

**Comment by mtibben** [comment]
Any further feedback on this @xabbuh @fabpot ?

**Comment by xabbuh** [comment]
I did not have the time to dig deeper, but I would prefer to first check why the `ServerRequest` class behaves this way and if this isn't a bug that needs to be fixed there.

**Comment by mtibben** [comment]
Any further feedback on this @xabbuh? I don't believe `ServerRequest` is the issue, the issue is `Request::getUri` (see https://github.com/symfony/psr-http-message-bridge/pull/30#discussion_r116972554)

**Comment by xabbuh** [comment]
I tried to debug this and IIRC the factory behaves differently depending on how you pass the request URI (through the constructor or with `withRequestTarget()`). Thus, I think this change looks indeed valid.

**Comment by fabpot** [comment]
Thank you @mtibben.
